### PR TITLE
AIP forkproperties

### PR DESCRIPTION
**Pull request by beidouz**
## Notice

It is not allowed to submit your PR to the master branch directly, please submit your PR to the master-pre-merge branch.

## Description

Please include a brief summary of the change that this pull request proposes. Include any relevant motivation and context. List any dependencies required for this change.

- add fork logic to getPrecompileContract

Fixes Issue # .

## Type of change

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [ ] Bug fix.
- [x] New feature.
- [ ] Enhancement.
- [x] Unit test.
- [x] Breaking change (a fix or feature that causes existing functionality to not work as expected).
- [ ] Requires documentation update.

## Testing

Please describe the tests you used to validate this pull request. Provide any relevant details for test configurations as well as any instructions to reproduce these results.

-

## Verification

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [x] I have self-reviewed my own code and conformed to the style guidelines of this project.
- [ ] New and existing tests pass locally with my changes.
- [x] I have added tests for my fix or feature.
- [ ] I have made appropriate changes to the corresponding documentation.
- [ ] My code generates no new warnings.
- [ ] Any dependent changes have been made.

### CODE REVIEW

File: modPrecompiled/src/org/aion/precompiled/ContractFactory.java [hunk 1]
```diff
@@ -78,30 +79,60 @@ public IPrecompiledContract getPrecompiledContract(
...
-            case ADDR_TOTAL_CURRENCY:
-            default:
-                return null;
+            }
+        } else {
+            switch (context.address().toString()) {
+                case ADDR_TOTAL_CURRENCY:
+                    // return new TotalCurrencyContract(track, context.sender(),
+                    // Address.wrap(OWNER));
+                    return null;
```
Comment (yulongaion) [inline modPrecompiled/src/org/aion/precompiled/ContractFactory.java]: Suggest wee keep the code before the fork the activated. If somebody calls it after the release and before the fork number, we will be in trouble.

File: modAionImpl/src/org/aion/zero/impl/config/CfgAion.java [hunk 2]
```diff
@@ -136,6 +139,51 @@ private void closeFileInputStream(final FileInputStream fis) {
         }
     }
 
+    public void setForkProperties() {
```
Comment (aion-kelvin) [inline modAionImpl/src/org/aion/zero/impl/config/CfgAion.java]: can we mention in JavaDoc that this sets it to mainnet?

File: modAionImpl/src/org/aion/zero/impl/config/CfgAion.java [hunk 3]
```diff
@@ -136,6 +139,51 @@ private void closeFileInputStream(final FileInputStream fis) {
...
+        setForkProperties("mainnet");
+    }
+
+    public void setForkProperties(String networkName) {
+        Properties properties = new Properties();
+
+        // old kernel doesn't support the fork feature.
+        if (networkName.equals("config")) {
```
Comment (aion-kelvin) [inline modAionImpl/src/org/aion/zero/impl/config/CfgAion.java]: Why is this needed?  Maybe I am missing something, but if this code is executing, doesn't  that mean they're using the new kernel? 

Or is this just to support the old kernel's config format?

File: modAionImpl/src/org/aion/zero/impl/config/CfgAion.java [hunk 4]
```diff
@@ -136,6 +139,51 @@ private void closeFileInputStream(final FileInputStream fis) {
...
+            return;
+        }
+
+        FileInputStream fis = null;
+        try {
```
Comment (aion-kelvin) [inline modAionImpl/src/org/aion/zero/impl/config/CfgAion.java]: optional: can use try-with-resources; i.e. `try (FileInputStream fis = new FileInputStream( ... )) { ... } ` and omit the `finally { /* close */ }`

File: modMcf/test/org/aion/mcf/config/CfgForkTest.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            new FileWriter(forkFile, true));
+        out.write("fork0.3.2=2000000");
+        out.close();
+
+        cfg = CfgAion.inst();
+    }
+
+    @After
+    public void teardown() {
+        forkFile.delete();
```
Comment (aion-kelvin) [inline modMcf/test/org/aion/mcf/config/CfgForkTest.java]: optional: I like to use `@TemporaryFolder` which deletes temp folders for you.  I find it makes the test code cleaner.  Since this PR needs to be released soon, we don't need to change it right now.  But maybe keep it in mind for the future.

File: modAionImpl/src/org/aion/zero/impl/config/CfgAion.java (hunk 3 above)
Comment (AionJayT) [reply modAionImpl/src/org/aion/zero/impl/config/CfgAion.java]: just for someone still uses old config struct to execute kernel.

File: modMcf/test/org/aion/mcf/config/CfgForkTest.java (hunk 5 above)
Comment (AionJayT) [reply modMcf/test/org/aion/mcf/config/CfgForkTest.java]: will put it as a ticket.

File: modAionImpl/src/org/aion/zero/impl/config/CfgAion.java (hunk 3 above)
Comment (aion-kelvin) [reply modAionImpl/src/org/aion/zero/impl/config/CfgAion.java]: Ok.  PR overall looks good to me.  Maybe consider putting a warning message for that case.

Comment (AionJayT) [reply modAionImpl/src/org/aion/zero/impl/config/CfgAion.java]: the reason doesn't print string here cause it will cause the packaging process failed.

File: script/prepack.sh [hunk 6]
```diff
@@ -47,7 +48,9 @@ fi
 # copy the config files if can't find the config env
 if [ ! -d "$CONFIG_PATH" ]; then
   mkdir $CONFIG_PATH
+  mkdir -p "${PACK_PATH}/aion/${DEFAULT_NETWORK}/config"
   cp -r ./modBoot/resource/** $CONFIG_PATH
+  cp  ${CONFIG_PATH}/${DEFAULT_NETWORK}/** "${PACK_PATH}/aion/${DEFAULT_NETWORK}/config"
```
Comment (AlexandraRoatis) [inline script/prepack.sh]: I don't understand the purpose of this change.

File: modPrecompiled/test/org/aion/precompiled/contracts/Blake2bHashTest.java [hunk 7]
```diff
@@ -24,15 +24,19 @@
...
 
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 import org.aion.base.util.ByteUtil;
-import org.aion.crypto.hash.Blake2b;
+import org.aion.mcf.config.CfgFork;
 import org.aion.precompiled.contracts.Blake2bHashContract;
```
Comment (AlexandraRoatis) [inline modPrecompiled/test/org/aion/precompiled/contracts/Blake2bHashTest.java]: has some unused imports

File: modPrecompiled/src/org/aion/precompiled/ContractFactory.java [hunk 8]
```diff
@@ -78,30 +80,59 @@ public IPrecompiledContract getPrecompiledContract(
...
+                                    track,
+                                    Address.wrap(ADDR_TOKEN_BRIDGE_INITIAL_OWNER),
+                                    Address.wrap(ADDR_TOKEN_BRIDGE));
+
+                    if (!context.origin().equals(Address.wrap(ADDR_TOKEN_BRIDGE_INITIAL_OWNER))
+                            && !contract.isInitialized()) {
+                        return null;
+                    }
+
+                    return contract;
```
Comment (AlexandraRoatis) [inline modPrecompiled/src/org/aion/precompiled/ContractFactory.java]: the `TokenBridgeContract` code could be moved to a method that can be called by both switch statements

File: modAionImpl/test/org/aion/zero/impl/cli/CliTest.java [hunk 9]
```diff
@@ -102,6 +102,11 @@
     private static final File mainnetGenesis = new File(MAIN_CONFIG_PATH, genesisFileName);
     private static final File testnetGenesis = new File(TEST_CONFIG_PATH, genesisFileName);
 
+    private static final File fork = new File(TEST_RESOURCE_DIR, forkFileName);
+    private static final File oldFork = new File(CONFIG_PATH, forkFileName);
```
Comment (AlexandraRoatis) [inline modAionImpl/test/org/aion/zero/impl/cli/CliTest.java]: need a test for behavior with old kernels config using the variable above

File: modAionImpl/test/org/aion/zero/impl/AionHubTest.java [hunk 10]
```diff
@@ -40,8 +40,10 @@
 import org.aion.zero.types.AionTransaction;
 import org.junit.After;
 import org.junit.BeforeClass;
+import org.junit.Ignore;
 import org.junit.Test;
 
+@Ignore
```
Comment (AlexandraRoatis) [inline modAionImpl/test/org/aion/zero/impl/AionHubTest.java]: why?

File: modAionImpl/src/org/aion/zero/impl/config/CfgAion.java [hunk 11]
```diff
@@ -136,6 +139,43 @@ private void closeFileInputStream(final FileInputStream fis) {
...
+            this.getFork().setProperties(properties);
+        } catch (Exception e) {
+            System.out.println(
+                    "<error on-parsing-fork-properties msg="
+                            + e.getLocalizedMessage()
+                            + ">, no protocol been updated.");
+        }
+    }
+
+    public void setForkProperties(String networkName) {
```
Comment (AlexandraRoatis) [inline modAionImpl/src/org/aion/zero/impl/config/CfgAion.java]: is this actually used?

File: script/prepack.sh (hunk 6 above)
Comment (AionJayT) [reply script/prepack.sh]: cause package the binary need to get the version number. If the cfgAion can't load the fork files. it will print something out. them the package will be mess up.

File: modAionImpl/test/org/aion/zero/impl/AionHubTest.java (hunk 10 above)
Comment (AionJayT) [reply modAionImpl/test/org/aion/zero/impl/AionHubTest.java]: accident commit.

File: modAionImpl/test/org/aion/zero/impl/cli/CliTest.java (hunk 9 above)
Comment (AionJayT) [reply modAionImpl/test/org/aion/zero/impl/cli/CliTest.java]: disable to support the old config compatibility for the fork. And will put it in the release note.

File: modAionImpl/src/org/aion/zero/impl/config/CfgAion.java (hunk 11 above)
Comment (AionJayT) [reply modAionImpl/src/org/aion/zero/impl/config/CfgAion.java]: removed
